docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,302 @@
+            if me.X() = pOther.X() and me.Y() = pOther.Y() then
+                return true
+            else
+                return false
+            end if
+        end method
+
+        method FormatAsString(in me) as string
+            return "(" & me.X() formatted as string & ", " & \
+                   me.Y() formatted as string & ")"
+        end method
+    end trait
+
+Traits can be derived from multiple supertraits.
+
+## Trait objects
+
+    public trait U
+        method type F(in me)
+    end trait
+
+    type T
+
+    implementation of U for T
+        method F(in me)
+        end method
+    end implementation
+
+    public handler G(in pParam as U)
+        pParam.F()
+    end handler
+
+A trait object is created whenever a concrete type `T` needs to be passed by
+trait `U` rather than directly as an `T`.  It packages the `T` value with
+information that identifies the methods of the implementation of `U` that was
+used.
+
+## Further work
+
+### Parameterized types
+
+Traits do not fully solve the problem of implementing `formatted as string`
+syntax with full generality and type safety.  There remains the problem of
+formatting polymorphic compound types such as `list`.  Consider the following
+naïve approach:
+
+    public implementation of CanFormatAsString for list
+        method FormatAsString(in me) as string
+            variable tString as string
+            put "(" into tString
+
+            variable tIter as any
+            repeat for each element tElement of me
+                variable tElement as CanFormatAsString
+                put tIter into tElement -- TYPE UNSAFE
+                put tElement formatted as string after tString
+            end repeat
+
+            put ")" after tString
+            return tString
+        end method
+    end implementation
+
+If the above method is compiled with strict type checking, then the marked line
+should fail to compile because `tIter` is of unknown type `any`, and therefore
+does not necessarily implement the `CanFormatAsString` trait.
+
+It would be better if the `CanFormatAsString` trait was only defined for lists
+where every element also conforms to `CanFormatAsString`.
+
+Similar arguments apply for the `array` type.
+
+To get full generality and type safety of the `formatted as string` trait
+Builder therefore also needs type specialization (e.g. `list of
+CanFormatAsString`).
+
+### Interaction with runtime typing
+
+    trait type T
+    end trait
+
+    implementation of T for integer
+    end implementation
+
+    handler Confuse(in pIn as T) as T
+        return pIn
+    end handler
+
+    public handler main()
+        variable tVar as any
+        put 15 into tVar        -- contents of tVar is an integer
+        Confuse(tVar) into tVar -- contents of tVar is a T trait object
+        add 1 to tVar           -- error?
+    end handler
+
+It is not clear what the correct behaviour is here.
```

Comment:
Note that this is analogous to an downcast in traditional OO systems.

    class X
        method f()
    class Y extends X
        method g()

    function h(X p) -> X

    var a -> X
    h(a).f()  -- Always
    h(a).g()  -- Downcast to Y is required.

In compile-time typing world, this shouldn't successfully compile.  In duck-typing world, there are no type restrictions at all, so this will work fine as long as `h()` returns *any* value with a `g()` method. We're somewhere in between.